### PR TITLE
Test deployer

### PR DESCRIPTION
**Pull request by bpostlethwaite**
@scjody this PR puts `bot` behind an Interface and uses it as a testing utility. It also mocks out `exec.Cmd` to test command execution behaviour. The interface pattern is pretty common but the mocking of the `exec.Run` is less so.  This post describes the pattern used here https://npf.io/2015/06/testing-exec-command/.

We are up to ~73% Coverage in `deployer/` and most of the remaining is Plugin Init specific that is hard to test without finer grained mocks of specific external dependencies.

I will probably add `go test` to CircleCI as a final step in this PR.

### CODE REVIEW

File: deployer/deployer.go [hunk 1]
```diff
@@ -95,7 +95,7 @@ func (dep *Deployer) ChatHandler(conv *plotbot.Conversation, msg *plotbot.Messag
 
 	if match := deployFormat.FindStringSubmatch(msg.Text); match != nil {
 		if dep.lockedBy != "" {
-			conv.Reply(msg, fmt.Sprintf("Deployment was locked by %s.  Unlock with '%s, unlock deployment' if they're OK with it.", dep.lockedBy, dep.bot.Config.Nickname))
+			conv.Reply(msg, fmt.Sprintf("Deployment was locked by %s.  Unlock with '%s, unlock deployment' if they're OK with it.", dep.lockedBy, dep.bot.Nickname()))
```
Comment (scjody) [inline deployer/deployer.go]: I realize this was broken in the original code, but this should be `AtMention()`

File: deployer/deployer.go [hunk 2]
```diff
@@ -144,10 +144,10 @@ func (dep *Deployer) ChatHandler(conv *plotbot.Conversation, msg *plotbot.Messag
 		bot.Notify(dep.config.AnnounceRoom, "#00ff00", fmt.Sprintf("%s has unlocked deployment", msg.FromUser.Name))
 	} else if msg.Contains("lock deploy") {
 		dep.lockedBy = msg.FromUser.Name
-		conv.Reply(msg, fmt.Sprintf("Deployment is now locked.  Unlock with '%s, unlock deployment' ASAP!", dep.bot.Config.Nickname))
+		conv.Reply(msg, fmt.Sprintf("Deployment is now locked.  Unlock with '%s, unlock deployment' ASAP!", dep.bot.Nickname()))
```
Comment (scjody) [inline deployer/deployer.go]: This too.

File: testutils/mockbot.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+func (bot *Bot) Nickname() string {
+	return bot.Config.Nickname
+}
+
+func (bot *Bot) AtMention() string {
+	return fmt.Sprintf("@%s:", bot.Myself.Name)
+}
+
+func (bot *Bot) WithMood(happy string, hyper string) string {
```
Comment (scjody) [inline testutils/mockbot.go]: Isn't this duplicating code from `mood.go`?

Comment (bpostlethwaite) [inline testutils/mockbot.go]: Yes it is required as deployer uses WithMood directly. Known problem with the plotbot plugin architecture: it's a leaky abstraction at best

Comment (scjody) [inline testutils/mockbot.go]: OK, fair enough.

**Comment by bpostlethwaite** [comment]
Yes it's a tradeoff for sure. I could probably reduce the delays by an order of magnitude but there may be a little less certainty in test consistency as the delay approaches the actual OS process start/stop time of the patched exec.Cmd operation.

Total test time is ~7 seconds which is high for a single Go test file but deployer should be an upper bound.

File: deployer/deployer.go (hunk 1 above)
Comment (bpostlethwaite) [inline deployer/deployer.go]: Yup I can make that change in this PR

Comment (bpostlethwaite) [inline deployer/deployer.go]: done in d47a9d8
That made it possible to remove the Nickname() interface method and associated implementations

**Comment by bpostlethwaite** [comment]
Okay @scjody plotbot testing in CI is working. I had to make a few more changes to get it working.

1. Fixup a few non deployer tests which failed when I ran repo wide tests
2. Drop `GoDep` as a dep mgmt tool. Instead we use `govendor` which uses the newer `vendor` model. We are vendoring all our assets now.
3. :hocho: a bunch of stuff in the README. It was outdated. I left some comments regarding `govendor` and flagged the Docker section as unmaintained.

**Comment by bpostlethwaite** [comment]
A last :+1: for these changes?

**Comment by scjody** [comment]
👍 💃

**Pull request merged**